### PR TITLE
[WIP] Auxiliary task similar to xent regularization

### PR DESCRIPTION
**Pull request by zhu-han**
Implement an auxiliary task described in https://github.com/k2-fsa/snowfall/issues/97#issue-805286306.

Tried two types of implementation. 

1. In addition to the MMI loss, we have two losses `xent_loss` and `smooth_loss`.
```
xent_loss = - (xent_output * num_post_mat).sum()
```
```
smooth_loss = (phone_counts * torch.log(phone_counts / phone_counts.sum() + 1.0e-20)).sum()
```
In this case, the regular output layer has three loss: MMI loss, `xent_loss` (since `num_post_mat` is from regular output layer) and `smooth_loss`.  And the auxiliary output layer has one loss: `xent_loss`.

2. Use `num_post_mat.detach()` to cut gradient from `xent_loss` to regular output layer. In this case,  I think `smooth_loss` may not be necessary. Therefore, the regular output layer has one loss: MMI loss. And the auxiliary output layer also has one loss: `xent_loss`.

Experiments are conducted with tdnn-lstm model. Haven't got a significant improvement yet.

- The baseline model is trained with MMI loss only. The model was trained three times. WERs are `10.39%`, `10.38%` and `10.11%`.

- Type 1 implementation results:

Configuration | WER
-- | --
xent_scale = 0.005   smooth_scale = 0.005 | 10.33%
xent_scale = 0.005 smooth_scale = 0.01 | 10.38%
xent_scale = 0.005 smooth_scale = 0.05 | 10.44%
xent_scale = 0.01 smooth_scale = 0.005 | 10.03%
xent_scale = 0.01 smooth_scale = 0.01 | 10.35%
xent_scale = 0.01 smooth_scale = 0.05 | 10.19%
xent_scale = 0.05 smooth_scale = 0.005 | 10.97%
xent_scale = 0.05 smooth_scale = 0.01 | 10.15%
xent_scale = 0.05 smooth_scale = 0.05 | 10.24%
xent_scale = 0.1 smooth_scale = 0.0 | 12.32%
xent_scale = 0.1 smooth_scale = 0.1 | 10.52%
xent_scale = 0.1 smooth_scale = 0.5 | 11.63%
xent_scale = 0.1 smooth_scale = 1.0 | 13.06%
xent_scale = 0.5 smooth_scale = 0.1 | 11.87%
xent_scale = 0.5 smooth_scale = 0.5 | 12.78%

, `xent_scale` and `smooth_scale` are loss scale of `xent_loss` and `smooth_loss` respectively.

- Type 2 implementation results:

Configuration | WER
-- | --
xent_scale = 0.005 | 10.14%
xent_scale = 0.01 | 10.45%
xent_scale = 0.05 | 10.13%
xent_scale = 0.1 | 10.23%
xent_scale = 0.3 | 10.81%
xent_scale = 0.5 | 10.32%
xent_scale = 1.0 | 11.87%

### CODE REVIEW

**Comment by danpovey** [comment]
Mm, that's a shame.
Could you try the type 2 loss with
xent_scale = 0.05
and the type 1 loss with
xent_scale = 0.05, smooth_scale = 0.05
?

**Comment by zhu-han** [comment]
Will try it.

**Comment by danpovey** [comment]
Also please make sure that the `num_post_mat.sum(0)` is mostly ones with a few zeros (where the final frames are).
Just in case there is some bug I didn't spot.

**Comment by zhu-han** [comment]
I think you means `num_post_mat.sum(1)`.  It is mostly ones with a few zeros as you said.

**Comment by zhu-han** [comment]
Added type 2 `xent_scale = 0.05`  result and type 1 `xent_scale = 0.05, smooth_scale = 0.05` result to the above tables. With smaller scale, these two results are better. I could try smaller scales to see if we can get better performance.

BTW, I got 9.89% WER with the baseline model before. The latest result I got is 10.39%.  So the randomness of training could make it hard for us to ensure the effectiveness of this new task. I wonder should we make the training process deterministic?

**Comment by danpovey** [comment]
I wonder if some other change in the code could have caused that difference in the baseline.  For example, I think we made a change in how the batching is done.  Perhaps try running the baseline once more?

**Comment by zhu-han** [comment]
Sure, will try running it again.

File: egs/librispeech/asr/simple_v1/mmi_xent_train.py [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+        cut_transforms=[
+            CutConcatenate(),
+            CutMix(
+                cuts=cuts_musan,
+                prob=0.5,
+                snr=(10, 20)
+            )
+        ]
+    )
+    train_sampler = SingleCutSampler(
```
Comment (pzelasko) [inline egs/librispeech/asr/simple_v1/mmi_xent_train.py:574]: @danpovey since he is using a `SingleCutSampler` and `CutConcatenate` a couple of lines before, I don't believe there are any significant changes to how the batches are formed w.r.t. the past experiments. For these to go into effect, you'd have to use the `BucketingSampler`.

Comment (danpovey) [reply egs/librispeech/asr/simple_v1/mmi_xent_train.py:574]: Oh OK. 
BTW, RE BucketingSampler: another concern I have is that if we use that, we'd have to do the same in test time because the model won't have "seen" data where there is a discontinuity.

**Comment by zhu-han** [comment]
Updated https://github.com/k2-fsa/snowfall/pull/100#issue-573949092 with some new results. 
(1) I run the baseline models two times again, get WER 10.38% and 10.11%.
(2) With smaller `xent_scale` and `smooth_scale`, still can't get significant improvement. These differences may just come from randomness.

**Comment by danpovey** [comment]
OK.  I feel we shouldn't give this up too readily because this is something that was consistently helpful in Kaldi.
Is it possible to have different versions of the final layer, e.g. the last lstm layer, for the regular vs. xent output?
I know this is a bit of a hassle, but I think it's worth doing the experiment.

**Comment by zhu-han** [comment]
Sure. Could try it.

**Comment by danpovey** [comment]
Can anyone make available for me one of the trained transformer models?
I want to have a look at the parameters..

**Comment by zhu-han** [comment]
This is a transformer model trained with MMI: [epoch-19.pt.zip](https://github.com/k2-fsa/snowfall/files/5994395/epoch-19.pt.zip)

**Comment by danpovey** [comment]
Thanks!

**Comment by danpovey** [comment]
Empty zip file :-(

**Comment by danpovey** [comment]
@zhu-han can you please check the zip file?

**Comment by zhu-han** [comment]
Sorry, this one should be right: https://drive.google.com/file/d/1luEAQbapIB0mH000JaonkTsAZppu4DWa/view?usp=sharing

**Comment by danpovey** [comment]
Thanks!

File: egs/librispeech/asr/simple_v1/mmi_xent_train.py (hunk 1 above)
Comment (pzelasko) [reply egs/librispeech/asr/simple_v1/mmi_xent_train.py:574]: I think that cut concatenation has always been disabled in test time (dev/test Dataset/Dataloader)

**Comment by danpovey** [comment]
@zhu-han would you mind sharing the epoch-1 and epoch-5 models  also?  I want to compare something..

**Comment by zhu-han** [comment]
@danpovey Uploaded to https://drive.google.com/drive/folders/1V5L8WaLDDdg5_oJl-GV8CMpdl6rg9FIL?usp=sharing

**Comment by pzelasko** [comment]
Regarding transformer: I noticed that "warmup steps" are 25000, isn't that a bit excessive? I think that would mean more than half of the training epochs are spent in the warmup mode... To my surprise Espresso/ESPnet recipes tend to stick to that value regardless of the size of the dataset, I wonder if this is not a relevant parameter, a magic value that "Just Works", or simply it's forgotten about?

**Comment by zhu-han** [comment]
@pzelasko I think smaller warmup steps may be better given current setup. I haven't tuned this yet, could try it later.

**Comment by zhu-han** [comment]
@danpovey I tried more experiments, such as the separate last lstm layer sturcture you mentioned before,  and a different training strategy: training the regular task for some epoch and then add the xent task for the remaining epochs. 

If we train the models for 20 epochs and decode after each epoch, we could find the mmi training without the xent task is consistently better than the one with xent task. The baseline mmi models could get WER around 9.8% between 12 ~ 20 epochs. However, with the xent task, WER is rarely below 10%.

I suspect the reason is the posteriors get from the regular output is less accurate than the one used in kaldi chain models, since no explicit frame level supervision is provided in snowfall's mmi training.
